src/components/hooks/useActivityStream.js

```diff
@@ -1,53 +1,129 @@
 }
 
 /**
  * Custom hook that returns activity data associated to the room of the given ID.
+ * If an element reference is given, it will attach an scroll listener to load more
+ * data as user scrolls to top.
  *
  * @param {string} roomID  ID of the room for which to return data.
- * @returns {Room} Activity ID associated to the room
+ * @param {object} elementRef  reference to the element to attach scroll listener.
+ * @param {object} lastElementRef  reference to the last element within elementRef
+ * @returns {Array} Activities state and showLoader state
  */
-export default function useActivityStream(roomID) {
-  const [activityIDs, dispatch] = useReducer(reducer, []);
+export default function useActivityStream(roomID, elementRef, lastElementRef) {
+  const [activities, dispatch] = useReducer(reducer, []);
+  const [showLoader, setShowLoader] = useState(false);
+  const [scrollToLastActivity, setScrollToLastActivity] = useState(true);
   const {roomsAdapter} = useContext(AdapterContext);
 
+  const loadPreviousActivities = async () => {
+    const previousActivities = await roomsAdapter.getPreviousRoomActivities(roomID);
+
+    dispatch({type: PREPEND_ACTIVITIES, payload: previousActivities});
+  };
+
+  const handleScroll = (event) => {
+    // Only load more data when user has reached the top
+    const isViewportTop = event.target.scrollTop === 0;
+
+    if (!showLoader && isViewportTop && roomsAdapter.hasMoreActivities(roomID)) {
+      setShowLoader(true);
+
+      // Throttle scroll
+      setTimeout(async () => {
+        await loadPreviousActivities();
+        setShowLoader(false);
+      }, 500);
+    }
+  };
+
+  // Keep requesting previous activities until activity stream is filled
   useEffect(() => {
-    const subscription = roomsAdapter.getRoomActivities(roomID).subscribe((activities) => {
-      dispatch({type: APPEND_ACTIVITIES, payload: activities});
+    const activityStreamBoundaries = elementRef.current.getBoundingClientRect();
+    const lastActivityBoundaries = lastElementRef.current.getBoundingClientRect();
+    const hasMoreActivities = roomsAdapter.hasMoreActivities(roomID);
+
+    if (lastActivityBoundaries.bottom < activityStreamBoundaries.bottom) {
+      // If there are more activities, load them
+      if (hasMoreActivities) {
+        loadPreviousActivities();
+        // if there are no more activities to load, scroll to last activity
+      } else {
+        lastElementRef.current.scrollIntoView();
+        setScrollToLastActivity(false);
+      }
+    } else {
+      // Scroll to bottom once after loading enough content
+      // eslint-disable-next-line no-lonely-if
+      if (scrollToLastActivity) {
+        lastElementRef.current.scrollIntoView();
+        setScrollToLastActivity(false);
+      }
+    }
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [activities]);
+
+  // Attach scroll event listener and subscribe to future updates on load
+  useEffect(() => {
```

Comment:
so two `useEffect`? elaborate a bit

Reply:
Because there are two different sets of side-effects:
1) Side effects to execute only once, on load
2) Side effects when there is a change to the activities